app/controllers/newflow/login_signup_controller.rb

```diff
@@ -0,0 +1,323 @@
+      @email_address = current_user.email_addresses.first&.value
+    end
+
+    def profile_newflow
+      render layout: 'application'
+    end
+
+    def reset_password_form
+      @email = login_failed_email
+    end
+
+    def reset_password
+      handle_with(
+        ResetPassword,
+        success: lambda {
+          @email = @handler_result.outputs.email
+          clear_newflow_state
+          sign_out!
+          security_log :reset_password_success, user: @handler_result.outputs.user
+          render :reset_password_email_sent
+        },
+        failure: lambda {
+          security_log :reset_password_failed, user: @handler_result.outputs.user
+          render :reset_password_form
+        }
+      )
+    end
+
+    def change_password_form
+      handle_with(
+        FindUserByToken,
+        success: lambda {
+          sign_in!(@handler_result.outputs.user)
+          security_log(:change_password_form_loaded)
+          render :change_password_form
+        },
+        failure: lambda {
+          security_log(:change_password_form_not_loaded, token: params[:token])
+          render status: 400
+        }
+      )
+    end
+
+    def change_password
+      # TODO: This check again here in case a long time elapsed between the GET and the POST
+      # user_signin_is_too_old?
+      # reauthenticate_user_if_signin_is_too_old!
+
+      handle_with(
+        ChangePassword,
+        user: current_user,
+        success: lambda {
+          security_log :password_reset
+          redirect_back(fallback_location: profile_newflow_url)
+        },
+        failure: lambda {
+          security_log :password_reset_failed
+          render :change_password_form, status: 400
+        }
+      )
+    end
+
+    def logout
+      sign_out!
+      redirect_back(fallback_location: newflow_login_path)
+    end
+
+    # Log in (or sign up and then log in) a user using a social (OAuth) provider
+    def oauth_callback
+      handle_with(
+        OauthCallback,
+        success: lambda {
+          authentication = @handler_result.outputs.authentication
+          user = @handler_result.outputs.user
+          unless user.is_activated?
+            # not activated means signup
+            save_unverified_user(user)
+            @first_name = user.first_name
+            @last_name = user.last_name
+            @email = @handler_result.outputs.email
+            security_log(:student_social_sign_up, user: user, authentication_id: authentication.id)
+            # must confirm their social info on signup
+            render :confirm_social_info_form and return
+          end
+          sign_in!(user)
+          # we should perhaps create a more specific security log
+          security_log(:student_authenticated_with_social, user: user, authentication_id: authentication.id)
+          redirect_back(fallback_location: profile_newflow_path)
+        },
+        failure: lambda {
+          # TODO: rate-limit this
```

Comment:
was this rate limited previously?

Reply:
Ohh, it doesn't look like it was. But I thought it should be, I may be wrong.

Removing that todo.